agora-test/Spec/Effect/TreasuryWithdrawal.hs

```diff
@@ -10,28 +10,272 @@ module Spec.Effect.TreasuryWithdrawal (tests) where
+definitionTWE = plam $ \input -> unTermCont $ do
+  datum <- tclet $ pfstBuiltin # input
+  txinfo <- tclet $ psndBuiltin # input
+
+  let scriptContext =
+        pcon $ PScriptContext $
+          pdcons @"txInfo" # pdata txinfo
+          #$ pdcons @"purpose" # pdata (pconstant $ Spending (TxOutRef "0b2086cbf8b6900f8cb65e012de4516cb66b5cb08a9aaba12a8b88be" 1))
+          # pdnil
+
+  pure $ opaqueToUnit #$ treasuryWithdrawalValidator currSymbol
+    # pforgetData (pdata datum)
+    # pforgetData (pdata (pconstant ()))
+    # scriptContext
+
+propertyTWE :: Property
+propertyTWE = classifiedProperty genTWECases shrinkTWE expectedTWE classifyTWE definitionTWE
+
+{- | Generates "lawful" ScriptContext from given TreasuryWithdrawalDatum.
+Other cases can use this ScriptContext to derive from and develop
+a case specific contexts with generators.
+
+TODO: will this work okay with generators adding and removing
+parts? I don't see particular reason it will not to, but will
+that be a "good" generator?
+-}
+txInfoFromTWEDatum :: (([TxInInfo], [TxOut]) -> Gen ([TxInInfo], [TxOut])) -> TreasuryWithdrawalDatum -> Gen TxInfo
+txInfoFromTWEDatum cb datum = do
+  (input, output) <- cb (expectedInput, expectedOutput)
+  return $
+      TxInfo
+      { txInfoInputs = input
+      , txInfoOutputs = output
+      , txInfoFee = Value.singleton "" "" 2
+      , txInfoMint = Value.singleton currSymbol validatorHashTN (-1)
+      , txInfoDCert = []
+      , txInfoWdrl = []
+      , txInfoValidRange = Interval.always
+      , txInfoSignatories = [signer]
+      , txInfoData = []
+      , txInfoId = "0b123412341234"
+      }
+  where
+    (expectedInput, excessOutputs)  = expectedTxInInfoFromTWEDatum datum
+    expectedOutput = expectedTxOutFromTWEDatum datum <> excessOutputs
+
+expectedTxOutFromTWEDatum :: TreasuryWithdrawalDatum -> [TxOut]
+expectedTxOutFromTWEDatum (TreasuryWithdrawalDatum r _) =
+  f <$> r -- add outputs to treasuries, returning excess STs.
+  where 
+    f (addr, val) = TxOut
+      { txOutAddress = Address addr Nothing
+      , txOutValue = val
+      , txOutDatumHash = Nothing
+      }
+
+{- | Generates expected inputs from given Datum
+-}
+expectedTxInInfoFromTWEDatum :: TreasuryWithdrawalDatum -> ([TxInInfo], [TxOut])
+expectedTxInInfoFromTWEDatum (TreasuryWithdrawalDatum r t) =
+  (inputGAT:((\addr -> TxInInfo
+    (TxOutRef "0b2086cbf8b6900f8cb65e012de4516cb66b5cb08a9aaba12a8b88be" 1)
+    TxOut
+    { txOutAddress = Address addr Nothing
+    , txOutValue = treasuryInputValue
+    , txOutDatumHash = Nothing
+    })
+  <$> t)
+  , [TxOut
+    { txOutAddress = Address (head t) Nothing
+    , txOutValue = extras
+    , txOutDatumHash = Nothing
+    }])
+  where
+    totalValues = mconcat $ snd <$> r
+    treasuryInputValue = distributeValue (length t) totalValues
+    extras = Value.unionWith (-) (mconcat (replicate (length t) treasuryInputValue)) $ totalValues
+
+distributeValue :: Int -> Value -> Value
```

Comment:
This definitely needs a doc string. It's not clear at all what this does from the type or the name.